src/libstd/sys/sgx/mod.rs

```diff
@@ -110,6 +110,50 @@ pub fn decode_error_kind(code: i32) -> ErrorKind {
     }
 }
 
+// This function makes an effort to sleep at least as long as `duration`.
+// Note that in general there is no guarantee about accuracy of time and
+// timeouts in SGX model. The enclave runner serving usercalls may lie about
+// current time and/or ignore timeout values.
+//
+// Once the event is observed, `stop` will be used to determine whether or not
+// we should continue to wait.
+//
+// FIXME: note these caveats in documentation of all public types that use this
+// function in their execution path.
+pub fn wait_timeout_sgx<F>(event_mask: u64, duration: crate::time::Duration, stop: F)
+where
+    F: Fn() -> bool,
+{
+    use self::abi::usercalls;
+    use crate::cmp;
+    use crate::io::ErrorKind;
+    use crate::time::Instant;
+
+    let start = Instant::now();
```

Comment:
Ok so basically you're using a deadline approach now. I thought we discussed this wasn't going to be necessary? Usercalls are slow and we should avoid doing them.

Reply:
Yeah, I initially though so too, but this seems necessary specially because spurious wakeups are possible. I think we can improve the perf by changing the parameter to wait to be a deadline instead.

Reply:
As discussed, please just re-use the timeout and some arbitrary decrement.

Reply:
As discussed:
```
let result = wait(as specified by user)
match result {
    timedout => return timedout
    non-spurious wakeup => return woken
    other => {
        loop {
            let result = wait(like before, but WAIT_NO)
            match result {
                non-spurious wakeup => return woken
                timedout (no event pending) => break
                other => (repeat)
            }
        }
        do the original wait, but use deadlines
    }
}
```